Adds LOG_*_ONCE macros

RafaelLaya opened the pull request:
Sometimes it can be useful to execute a log statement only a fixed number of times. For instance, because logging it would produce a lot of spam and it's a fact that will stay constant during the whole runtime. This change adds macros to silence a single log statement after a given number of executions. 

Signed-off-by: Rafael Laya <rafaellaya@fb.com>

nordic-krch reviewed (changes requested):
Please add test to `tests/subsys/logging/log_api`. If macro for limited number of executions is added to sys/util add also test there.

RafaelLaya commented on include/logging/log.h:
```diff
@@ -39,6 +39,17 @@ extern "C" {
+ * @param ... A string optionally containing printk valid conversion specifier,
+ * followed by as many values as specifiers.
+ */
+#define LOG_ERR_ONCE(...)    Z_LOG_ONCE(LOG_LEVEL_ERR, __VA_ARGS__)
```
Great suggestion!

RafaelLaya replied:
Hi again @nordic-krch, I've made some local changes and about to make tests. However, I might be missing how to run the tests properly. For instance:
1. I go to ```zephyr/tests/subsys/logging/log_api/src/test.inc``` 
2. As an attempt to break ```test_log_various_messages()```, I replace the LOG_DBG() in line 125 by some garbage like ```LOG_DBG("garbage");```
3. Save
4. Then from zephyr root I run ```./scripts/twister -T tests/subsys/logging/log_api```
5. I see the tests passing even though I tried to break one

Did I miss something about running the tests?

nordic-krch replied:
you're right. It got broken when new ztest got introduced. You can try with fix: #41872.

RafaelLaya replied:
Thanks! that fix helped

RafaelLaya commented on include/logging/log_core.h:
```diff
@@ -357,9 +357,21 @@ static inline char z_log_minimal_level_to_char(int level)
 
+#define Z_LOG_DO_ACTION_ONCE(code)	\
+({ \
+	static _bool done_once = false; \
```
Sorry that must be a typo

dcpleung replied:
Might consider changing the name to something like `log_done_once` so it will be less confusing when someone is going through the symbol table.

RafaelLaya replied:
Thanks! Indeed it was a confusing name. I moved to sys_utils (since it's not log-specific) and renamed as ```DO_FIRST_N``` so that it's more clear that it just silences a piece of code after a given ```n```. Does that sound better?

RafaelLaya commented:
- Re-named ```Z_LOG_DO_ACTION_ONCE``` to ```DO_FIRST_N``` (```DO_FIRST_N(1, ...)```) is equivalent to ```Z_LOG_DO_ACTION_ONCE```) and moved to sys/util_macro.h
- Added ```DO_FIRST_N``` tests to ```tests/unit/util/test.inc```
- Re-named ```LOG_*_ONCE to LOG_*_N``` and added parameter ```n``` 
- Added tests to ```tests/subsys/logging/log_api/src/test.inc```

RafaelLaya commented:
Fixed tests with CONFIG_LOG_FUNC_NAME_PREFIX_DBG=y and CONFIG_SAMPLE_MODULE_LOG_LEVEL_DBG=y

RafaelLaya commented:
Split original into two commits

nordic-krch commented on include/logging/log.h:
```diff
@@ -39,6 +39,17 @@ extern "C" {
 #define LOG_ERR(...)    Z_LOG(LOG_LEVEL_ERR, __VA_ARGS__)
 
+/**
+ * @brief Writes an ERROR level message to the log, but only the first @p n times
```
English is not my primary language but in zephyr we usual don't use third person verb.
`Write an ERROR ...`

RafaelLaya replied:
I copied that portion of the docstring from the other macros and it was already in third person. However this reminds me to add the "@param n" documentation to the docstring

nordic-krch replied:
yeah, you're right. This header was one of my first contributions to zephyr and i did that mistake.

nordic-krch reviewed (commented):
Codewise looks good. Please remove `[]` from commit messages. Also split commits into testing part and feature part.

RafaelLaya committed:
utils: add macro for silencing code after N executions

Sometimes you might want to execute a piece of code and then silence it.
This macro allows you to do so.

Signed-off-by: Rafael Laya <rafaellaya@fb.com>

RafaelLaya committed:
tests: add tests to util for macro that silences code after N executions

This add tests to the macro DO_FIRST_N() which is used to silence a
piece of code after it has executed N times.

Signed-off-by: Rafael Laya <rafaellaya@fb.com>

RafaelLaya committed:
logging: Fix verbs in log.h for consistency with the rest of the codebase

Zephyr does not use 3rd person verb in docstrings. This fixes the verbs
for log.h.

Signed-off-by: Rafael Laya <rafaellaya@fb.com>

RafaelLaya committed:
logging: Implement macros for logs with limited execution

Adds macros to silence a log after N executions, using the macro
DO_FIRST_N()

Signed-off-by: Rafael Laya <rafaellaya@fb.com>

RafaelLaya committed:
tests: Add tests for logging macros that have limited execution

This adds tests to the logging macros that silence after N executions.

Signed-off-by: Rafael Laya <rafaellaya@fb.com>

RafaelLaya commented:
Split into: 5 commits
1. add macro for silencing code after N executions
2. add tests to macro that silences code after N executions
3. Fix verbs in log.h for consistency with the rest of the code-base
4. Implement macros for logs with limited execution
5. Add tests for logging macros that have limited execution

nordic-krch commented:
Commits with test changes should start with `tests: `.

RafaelLaya commented:
* Re-named commits that add tests so that they start with ```tests: ```
* Fixed missing parameters in doxygen documentation
* Fixed white-space in DO_FIRST_N()
* Added Signed-off-by in the body of the commits

Hopefully that fixes the CI failures!

RafaelLaya commented:
- Fixed the doxygen which was the single last CI test that was failing. Ran the Document Generation engine locally and it ran successfully so should be fine now.

RafaelLaya commented:
* Fixed the author of the commit so it matches the signed-off-by line

nordic-krch reviewed (dismissed):
Looks good, please fix the sign-off email address.

RafaelLaya commented:
Should be fixed now! I learned about gitlint so I checked locally that there are no errors

nordic-krch commented:
Still not fixed. You are pushing from`Rafael Laya <rafaellaya@fb.com>` and sign off with `Rafael Laya Alfonzo <rafaellaya@fb.com>`

RafaelLaya commented:
I see, in that case I've removed the Alfonzo part from the signed-off-by line

nordic-krch commented:
CI is green but it won't be merged until v3.0.0 comes out as there is a feature freeze phase now.

RafaelLaya commented:
Great! thanks for the patience during my first contribution to Zephyr :)

nordic-krch commented:
> Great! thanks for the patience during my first contribution to Zephyr :)

No problem, I hope there will be more :)

RafaelLaya commented:
Resolved merge conflict

nordic-krch commented:
@RafaelLaya please resolve conflict. We could merge it soon.

carlescufi commented:
@RafaelLaya please rebase

bbilas commented:
@RafaelLaya would you mind re-opening that MR and preparing it to be merged? That's a super helpful feature that I want to see in the upstream.
